staging/src/k8s.io/client-go/rest/request_watchlist_test.go

```diff
@@ -0,0 +1,337 @@
+			}
+			if !fakeWatcher.IsStopped() {
+				t.Fatalf("the watcher wasn't stopped")
+			}
+		})
+	}
+}
+
+func TestWatchListFailure(t *testing.T) {
+	scenarios := []struct {
+		name        string
+		ctx         context.Context
+		watcher     *watch.FakeWatcher
+		watchEvents []watch.Event
+
+		expectedError error
+	}{
+		{
+			name: "request stop",
+			ctx: func() context.Context {
+				ctx, ctxCancel := context.WithCancel(context.TODO())
+				ctxCancel()
+				return ctx
+			}(),
+			expectedError: fmt.Errorf("context canceled"),
+		},
+		{
+			name: "stop watcher",
+			watcher: func() *watch.FakeWatcher {
+				w := watch.NewFake()
+				w.Stop()
+				return w
+			}(),
+			expectedError: fmt.Errorf("unexpected watch close"),
+		},
+		{
+			name:          "stop on watch.Error",
+			watchEvents:   []watch.Event{{Type: watch.Error, Object: &apierrors.NewInternalError(fmt.Errorf("dummy errror")).ErrStatus}},
+			expectedError: fmt.Errorf("Internal error occurred: dummy errror"),
+		},
+		{
+			name:          "incorrect watch type (Deleted)",
+			watchEvents:   []watch.Event{{Type: watch.Deleted, Object: makePod(1)}},
+			expectedError: fmt.Errorf("unexpected watch event .*, expected to only receive watch.Added and watch.Bookmark events"),
+		},
+		{
+			name:          "incorrect watch type (Modified)",
+			watchEvents:   []watch.Event{{Type: watch.Modified, Object: makePod(1)}},
+			expectedError: fmt.Errorf("unexpected watch event .*, expected to only receive watch.Added and watch.Bookmark events"),
+		},
+		{
+			name:          "unordered input returns an error",
+			watchEvents:   []watch.Event{{Type: watch.Added, Object: makePod(3)}, {Type: watch.Added, Object: makePod(1)}},
+			expectedError: fmt.Errorf("cannot add the obj .* with the key = ns/pod-1, as it violates the ordering guarantees provided by the watchlist feature in beta phase, lastInsertedKey was = ns/pod-3"),
+		},
+	}
+
+	for _, scenario := range scenarios {
+		t.Run(scenario.name, func(t *testing.T) {
+			target := &Request{}
+			if scenario.watcher == nil {
+				scenario.watcher = watch.NewFake()
+			}
+			if scenario.ctx == nil {
+				scenario.ctx = context.TODO()
+			}
+			go func(w *watch.FakeWatcher, watchEvents []watch.Event) {
+				for _, event := range watchEvents {
+					w.Action(event.Type, event.Object)
+				}
+			}(scenario.watcher, scenario.watchEvents)
+
+			res := target.handleWatchList(scenario.ctx, scenario.watcher)
+			resErr := res.Into(nil)
+			if resErr == nil {
+				t.Fatal("expected to get an error")
+			}
+			matched, err := regexp.MatchString(scenario.expectedError.Error(), resErr.Error())
+			if err != nil {
+				t.Fatal(err)
+			}
+			if !matched {
+				t.Fatalf("unexpected err = %v, expected = %v", resErr, scenario.expectedError)
+			}
+			if !scenario.watcher.IsStopped() {
+				t.Fatalf("the watcher wasn't stopped")
+			}
+		})
+	}
+}
+
+func TestWatchListWhenFeatureGateDisabled(t *testing.T) {
```

Comment:
Can we remove it and add that in a separate PR once the PR mentioned is merged?

Reply:
sure, np.